### PR TITLE
JSON encoding

### PR DESCRIPTION
**Pull request by mattjbray**
Encode goals and results to JSON, for display in imandra-web.

### CODE REVIEW

File: src-alcotest/imandra_goals_alcotest.ml [hunk 1]
```diff
@@ -75,6 +75,21 @@ let run_tests ?report_name () =
...
+      let write_report () =
+        CCFormat.printf "Writing JSON to %s/%s.html@." (Sys.getcwd ())
+          json_fname;
+        let module E : Decoders.Encode.S = Decoders_yojson.Basic.Encode in
+        let module Encode = Imandra_goals.Encode (E) in
+        let goals = Imandra_goals.all () in
+        let json_str = E.encode_string Encode.goals goals in
+        CCIO.with_out json_fname (fun oc -> output_string oc json_str)
+      in
+      at_exit write_report
```
Comment (c-cube) [inline src-alcotest/imandra_goals_alcotest.ml:91]: can we turn this all into a:

```ocaml
let with_json_dump k =
  Fun.protect ~finally:(fun () -> write_to_json_file ()) @@ fun () ->
  (* rest of program *)
  …
```
?

**Review by c-cube — APPROVED** [review]
LGTM :+1: 

had some minor quibble about `at_exit` but it might be needed for tests. No worries.

File: src-alcotest/imandra_goals_alcotest.ml (hunk 1 above)
Comment (mattjbray) [reply src-alcotest/imandra_goals_alcotest.ml:91]: Hmm I refactored to this https://github.com/imandra-ai/imandra-goals/pull/12/commits/a5839ba05a411b388b0109046ecc54b2659f52e0 but then the reports don't get written. Does `finally` get called if someone else (alcotest) calls `exit 0`?

Comment (c-cube) [reply src-alcotest/imandra_goals_alcotest.ml:91]: no, it directly jumps to the at_exit callbacks and  then terminates.

**Pull request merged**